Apodini Authorization

bauer-andreas opened the pull request:
# Apodini Authorization

## :recycle: Current situation & Problem

Apodini currently requires manual effort to do any kind of authentication or authorization.

## :bulb: Proposed solution

With this PR the collection Apodini framework components adds the ability to easily declare authentication and authorization
requirements on `Component`s.

A basic `Authorize` Metadata Declaration might then look like in the `Handler` shown below. 
Every authentication and authorization step consists of the following parts:
- The `Authenticatable` type representing the authenticated and authorized state.
- The `AuthenticationScheme` that is used to perform the user or protocol defined authentication scheme (while `BasicAuthenticationScheme` and `BearerAuthenticationScheme` are supported out of the box).
- A `AuthenticationVerifier` instance, which maps the output of the authentication scheme to the `Authenticatable` instance, and performs some initial verifications steps (e.g. password checks or signature checks).
- The Block of Authentication Requirements. These are evaluated against the `Authenticatable` instance. Predefined Authentication Requirements are `Allow`, `Deny` and `Verify`.

```swift
struct ExampleHandler: Handler {
    // ...
    var metadata: Metadata {
        Authorize(SomeAuthenticatable.self, using: SomeAuthenticationScheme(), verifiedBy: SomeAuthenticationVerifier()) {
            Verify(if: \.someProperty == "SomeValue")
            // ... `AuthorizationRequirement`s are listed here
        }
    }
}
```

The PR introduces out of the box support for **JSON Web Tokens** using the Vapoe `JWTKit`.

## :gear: Release Notes 

We introduce the new targets `ApodiniAuthorization` together with out of the box support for Basic Authentication (`ApodiniAuthorizationBasicScheme` target), Bearer Authentication (`ApodiniAuthorizationBearerScheme` target) and **Json Web Tokens** (`ApodiniAuthorizationJWT` target).

While the package provides the flexibility to support any arbitrary authentication and authorization mechanism, we want to specifically highlight JWT support. Refer to the documentation on how to define custom authentication schemes, verifiers and Authenticatables.

As the initial step, configure the `JWTSigner` in your web service as follows:
```swift
import Apodini
import ApodiniAuthorizationJWT

struct ExampleWebService: WebService {
    var content: some Component {
        // ...
    }

    var configuration: Configuration {
        JWTSigner(.hs256(key: "secret"))
    }
}
```

For a more detailed overview of signer configuration have a look at the [JWTKit Documentation](https://github.com/vapor/jwt-kit).

The next step is to define your Token type, by conforming to `JWTAuthenticatable`:
```swift
struct SomeToken: JWTAuthenticatable {
    var exp: ExpirationClaim

    var isAdmin: Bool
}
```
The `JWTAuthenticatable` supports any JWT claims as defined by the [JWTKit](https://github.com/vapor/jwt-kit).

As the last step is to declare the `Authorize` Metadata on all the Apodini `Component`s you want to be access protected.
Below is a example `Handler` which does three things:
- Declares a `Authorize` Metadata which requires an authenticated and authorized `SomeToken` (authenticated via Bearer Authentication Scheme).
- Specifies the `Verify(notExpired:)` authorization requirement to verify the `ExpirationClaim`.
- Declares the `Authorized` property to get access to the `SomeToken` instance in the handle method.
```swift
struct ExampleHandler: Handler {
    var token = Authorized<SomeToken>()

    func handle() throws -> String {
        let tokenInstance = try token()
        // ...
        return ...
    }

    var metadata: Metadata {
        Authorize(SomeToken.self) {
            Verify(notExpired: \.exp)
        }
    }
}
```

## :heavy_plus_sign: Additional Information

This PR introduces the `ApodiniHTTPProtocol`, decoupling the HTTP Information instances from the VaporSupport package.

Some parts of the API shall gain `async` capabilities in the future (e.g. the `AuthenticationVerifier` to support database operations and other async verification processes). My plan would be to do that in a follow up PR (or wait if #317 is potentially merged first). I didn't want to get any EventLoopFuture ugliness into here and introduce some sort of hybrid.
But feel free to leave any feedback.

### Related PRs
- #310 introduce the necessary infrastructure to add DelegatingHandlerInitializer from Metadata

### Testing
This PR adds a new `ApodiniAuthorizationTests` target containing extensive testing for the introduced features.

### Reviewer Nudging

The code includes extensive documentation and examples. You might want to start by looking at the test cases, maybe specifically at the `JWTTests` which give real world examples.
Further, the main `AuthorizationMetadata` and `OptionalAuthorizationMetadata` provide overview of how each components interact with each other.

bauer-andreas committed:
Create new ApodiniAuthorization feature

bauer-andreas committed:
Fix InformatioSet.insert()

bauer-andreas committed:
Introduce a dedicated, generalised Verify AuthorizationRequirement

bauer-andreas committed:
Authorize will evaluate requirements if found a existing authoirzed instance by default.
Completed JWT Tests

bauer-andreas committed:
Rework environment value handler and added extensive testing.

bauer-andreas committed:
Finalize documentation

bauer-andreas committed:
Added test vector and simplified environment storage

bauer-andreas committed:
Remove forgotten todo

bauer-andreas committed:
Remove the fluent auth package

theMomax commented on Package.swift:
```diff
@@ -78,7 +86,9 @@ let package = Package(
-        .package(url: "https://github.com/soto-project/soto-s3-file-transfer", from: "0.3.0")
+        .package(url: "https://github.com/soto-project/soto-s3-file-transfer", from: "0.3.0"),
+
+        .package(url: "https://github.com/vapor/jwt-kit.git", from: "4.0.0")
```
maybe add a `// Authorization` so its separated from `// Deploy`

theMomax commented on Sources/Apodini/Information/AnyInformation.swift:
```diff
@@ -10,6 +10,17 @@ public protocol AnyInformation {
+
+    /// Returns the untyped version of the ``AnyInformation`` instance. This method has no effect on ``Information``
+    /// instances and returns self. For ``InformationInstantiatable`` it returns the corresponding ``Information`` instance.
+    /// - Returns: The untyped ``AnyInformation``.
```
Just a general comment. The correct DocC syntax would be to leave a blank line between the "Definition" and the "Discussion" like this:

```swift
/// Returns the untyped version of the ``AnyInformation`` instance.
/// 
/// This method has no effect on ``Information`` instances and returns self. For
/// ``InformationInstantiatable`` it returns the corresponding ``Information`` instance.
/// 
/// - Returns: The untyped ``AnyInformation``.
```

theMomax commented on Sources/Apodini/Information/InformationSet.swift:
```diff
@@ -5,21 +5,37 @@
+    /// Pretty much acts a typed cache for the above storage dict.
+    /// For every key contained in this dictionary, there is a according untyped value preset above.
+    /// The other way around doesn't hold.
+    private var capturedInstantiatables: [AnyHashable: AnyInformation] = [:]
```
I think it would be a good idea to add a layer of abstraction here. I smell quite some bug potential🙃. Introduce a new wrapper around both `storage` and `capturedInstantiables` that abstracts the caching away please.

bauer-andreas replied:
I thought about the same. Though I kept it this way as it avoid to introduce unnecessary mappings and inefficiency for the below Collection and Sequence conformance. As you can see, currently we can just map the according protocol 1:1 to the Collection and Sequence requirements.
After all those two properties are completely private and never see the outside of an `InformationSet`. And everything is thorough covered by test vectors.
So, I would argue against introducing another level of abstraction.

theMomax replied:
If the caching mechanism is covered by regression tests, I am fine with leaving it as is.

theMomax commented on Sources/Apodini/Response/ApodiniError.swift:
```diff
@@ -83,17 +83,42 @@ public struct ApodiniError: Error {
+        return ApodiniError(
+            type: type,
+            reason: reason ?? error.reason,
+            description: description ?? error.description,
```
Shouldn't `error` have precedence over `self` (at least that holds for `information.merge`, doesn't it)?

theMomax replied:
One could also add an "original error" part to the reason/description...

bauer-andreas replied:
you are completely right. Great catch 👍

bauer-andreas replied:
> One could also add an "original error" part to the reason/description...

How do you envision that?

theMomax replied:
basically `reason: "\(reason) (original reason: \(error.reason))"` and the same for description.

theMomax commented on Sources/ApodiniAuthorization/Authenticatable.swift:
````diff
@@ -0,0 +1,6 @@
+// Created by Andreas Bauer on 07.07.21.
+//
+
+/// Any type conforming to ```Authenticatable`` represents some state of a authenticatable instance.
````
"an authenticatable"...and I am not sure what exactly you mean by that. Maybe add a discussion in the form of "An ``Authenticatable`` can be used together with ``?`` to do/create ?.

bauer-andreas replied:
I'll try to clarify that by adding further documentation 👍

theMomax commented on Sources/ApodiniAuthorization/AuthorizationErrorReason.swift:
```diff
@@ -0,0 +1,34 @@
+    case failedAuthorization
+    /// Represents some custom, potentially user defined error reason, to be used if the others don't fit.
+    /// A string based reason shall be supplied for debugging uses.
+    case custom(_ reason: String = "undefined")
```
I'd remove the default "undefined".

theMomax commented on Sources/ApodiniAuthorization/AuthorizationRequirements/AnyAuthorizationRequirement.swift:
```diff
@@ -0,0 +1,30 @@
+// Created by Andreas Bauer on 09.07.21.
+//
+
+/// Represents a type erase ``AuthorizationRequirement``
```
erased

theMomax commented on Sources/ApodiniAuthorization/AuthorizationRequirements/AnyAuthorizationRequirement.swift:
```diff
@@ -0,0 +1,30 @@
+// MARK: AnyAuthorizationRequirement
+public extension AuthorizationRequirement {
+    /// Default implementation calling the typed evalute version.
+    func anyEvaluate<Element0: Authenticatable>(for element: Element0) throws -> RequirementResult {
```
Why Element**0**? Just seems kinda weird...

bauer-andreas replied:
`Element` is already defined in the `AuthorizationRequirement` and I went the classic way of just adding an index.
But I can try to envision some more creative naming.

theMomax replied:
Ok...if that's the classic way...I'd have made it an `E` ;)

theMomax commented on Sources/ApodiniAuthorization/AuthorizationRequirements/AuthorizationCondition.swift:
````diff
@@ -0,0 +1,114 @@
+/// condition = \.someString != <expression>
+/// ```
+///
+/// ## Comparing Operators
````
Comparison

theMomax commented on Sources/ApodiniAuthorization/AuthorizationRequirements/AuthorizationCondition.swift:
````diff
@@ -0,0 +1,114 @@
+/// ```
+///
+/// ## Logical Operators
+/// The typical `&&`, `||` and `!` logical operators cna be used on ``AuthorizationCondition`` instances.
````
can

theMomax commented on Sources/ApodiniAuthorization/AuthorizationRequirements/AuthorizationRequirements.swift:
```diff
@@ -0,0 +1,47 @@
+
+    /// Initializes a new ``AuthorizationRequirements`` instance for a single ``AuthorizationRequirement``.
+    /// - Parameter requirement: The ``AuthorizationRequirement`` to be wrapped.
+    public init<Requirement: AuthorizationRequirement>(_ requirement: Requirement) where Requirement.Element == Element {
```
Why does it only take one element? I expected an empty initializer accompanied by a
```swift
public func with<Requirement: AuthorizationRequirement>(_ requirement: Requirement) -> Self where Requirement.Element == Element
```

theMomax replied:
Ah...you're using a resultBuilder...does it need to be public then?

bauer-andreas replied:
Yes, it is required to be public as it is used as a default value for an initializer parameter in the `AuthorizationMetadata` and `OptionalAuthorizationMetadata`.

theMomax commented on Sources/ApodiniAuthorization/AuthorizedProperty.swift:
```diff
@@ -0,0 +1,115 @@
+private extension Application {
+    class WrappedContainer {
+        var stateContainer = AuthorizationStateContainer()
+    }
```
ApodiniUtils defines `Box` for such purposes.

bauer-andreas committed:
Allow the result of the Authorized property to be discarded.
That allows to use the Authorized proprety without actually processing the result,
which specifically is usefule for a fail-safe, to ensure that any Security Metadata actually is translated into the corresponding enforcement.

theMomax commented on Sources/ApodiniAuthorization/AuthorizedProperty.swift:
```diff
@@ -0,0 +1,115 @@
+    }
+}
+
+/// The ``SomeAuthorizationStateContainer`` type is a type erasing wrapper around an ``AuthorizationStateContainer``
```
Is this an outdated comment?

bauer-andreas replied:
yes indeed 😁

theMomax commented on Sources/ApodiniAuthorization/AuthorizedProperty.swift:
```diff
@@ -0,0 +1,115 @@
+
+    var wrapped: WrappedContainer {
+        WrappedContainer()
+    }
```
`Delegate.environment(_:_:)` only changes the **local** environment. You thus don't have to implement a storage, if you don't want to insert global default values or anything. If you would use those, I think the current implementation would even share data across different client-connections. Thus you should be able to just do the following:

```swift
private extension Application {
    var stateContainer: AuthorizationStateContainer {
        get {
            fatalError()
        }
        nonmutating set {
            fatalError()
        }
    }
}

```

bauer-andreas replied:
The reason why I didn't just went with an `@EnvironemtObject` is that an `@Enviroment` property pulls out the value from the application key path if there isn't a value in the **local** environment.
This is therefore required to ensure that there is **always** a value present in the KeyPath, such that the Handler doesn't crash if something failed.

So using a fatalError in the `get` is definitely not possible, but I could probably use the `nonmutating set` to remove the need of the wrapper, as far as I understand that (as `.environment` requires an `WriteableKeyPath`)

> I think the current implementation would even share data across different client-connections

It doesn't. Its basically a copy-on-write. The first time the `@Environment` property is used it uses the statically, **empty** defined AuthorizationStateContainer from the extension. If an authenticator adds new state, it copies the value, modifies it accordingly (e.g. inserting the authorized Authenticatable instance) and inserts it into the **local** environment of the respective handler evaluation.

theMomax commented on Sources/ApodiniAuthorization/Delegates/AuthenticationScheme.swift:
````diff
@@ -0,0 +1,57 @@
+    /// ```
+    ///
+    /// An ``AuthenticationScheme`` might also define custom `ApodiniError` options which can be parsed here,
+    /// inorder to support wire format specific features. An error containing such custom error options
````
in order

theMomax commented on Sources/ApodiniAuthorization/Delegates/AuthenticationVerifier.swift:
```diff
@@ -0,0 +1,28 @@
+    associatedtype Element: Authenticatable
+
+    /// This method uses the ``AuthenticationInfo`` result received from the respective ``AuthenticationScheme``
+    /// to instantiate the ``Element`` instance. The instantiating process might includes calls to databases
```
include

theMomax commented on Sources/ApodiniAuthorization/Delegates/Authenticator.swift:
```diff
@@ -0,0 +1,130 @@
+        case let .fulfilled(cause), let .undecided(cause): // undecided is a acceptance state as well!
+            logger.trace("Authorization on Handler \(H.self) succeeded with \(cause())")
+            if authenticatable.isAuthorized {
+                return try delegate().handle()
```
Why don't you manually set the `instance` here as in L108 and the other cases? I am not sure what the effects are and why there should be a difference. I also think that if you don't override it, the `delegate` automatically uses the same `AuthorizationStateContainer` as  used by the local `authenticatable`.

bauer-andreas replied:
The `evaluateRequirements(...)` method is not only called once an instance is fully authorized, but also at the beginning of `handle()` if the authenticator finds that there is already an authorized instance (of the same `Authenticatable` type).
See:
https://github.com/Apodini/Apodini/blob/37de2a6567d6e3190764c9d397728409ed9faef3/Sources/ApodiniAuthorization/Delegates/Authenticator.swift#L38-L45

If an authenticator finds that there is already an authorized instance, it skips its own authentication and authorization operations. Only if not turned off via `skipRequirementsForAuthorized` (which is false by default), it will evaluate the authorization requirements against the existing instance.
And that's the case what you commented above. If there is already an existing instance, and the requirements are fulfilled, there is no reason to change something in the environment, because after all nothing changed.

theMomax commented on Sources/ApodiniAuthorization/AuthorizationRequirements/RequirementResult.swift:
```diff
@@ -0,0 +1,41 @@
+    case fulfilled(cause: @autoclosure () -> Cause = .unspecified)
+    /// Represents a neutral ``RequirementResult``.
+    /// Captures the ``Cause`` leading to the result. The ``Cause`` is only captured for debug purposes.
+    case undecided(cause: @autoclosure () -> Cause = .unspecified)
```
`Authenticator` interprets this case as successful. I don't think that is obvious. Please add an according note to the docs.

bauer-andreas replied:
I'll add some docs 👍

theMomax reviewed (commented):
Looks pretty solid👍🏻 I mostly focused on the generic implementation (`Delegate`/`Metadata` related stuff) and less on the correctness of the middleware-specific implementation details.

I had some comments and questions here and there. I'll go over the test cases and complete the review once those are resolved.

bauer-andreas committed:
Small adjustments

bauer-andreas committed:
Incorporate feedback

bauer-andreas committed:
Make AuthenticationScheme and AuthenticationVerifier async

bauer-andreas committed:
Ensure REUSE compliance

Commit:
Add support for arbitrary Metadata Blocks in Delegates.
This adds support for e.g. declaring Metadata on AuthenticationScheme

Commit:
Address linting issues

Commit:
Add .idea folder to .gitignore

bauer-andreas commented:
boschbruegge.ase.in.tum.de agent is out of space, therefore the last few bamboo builds are failing.

theMomax commented on Sources/Apodini/Response/ApodiniError.swift:
```diff
@@ -87,14 +107,26 @@ public struct ApodiniError: Error {
+        guard let new = newMaybe, let previous = previousMaybe else {
+            return newMaybe ?? previousMaybe
+        }
+        return "\(new) (original \(name): \(previous)"
```
I think that is the wrong way round. Usually, you would have the following usage:

```swift
struct SomeHandler: Handler {
    @Throws(reason: "Authorization failed") var authError

    var authorizedUser: Authorized<User>

    func handle() throws {
        do {
            let x = try authorizedUser()
        } catch {
            throw authError.merge(with: error.apodiniError)
        }
    }
}
```

With the current code you'd get: `Invalid Credentials (original reason: Authorization failed)`. At least that is how I'd use it. Of course you could argue that the call should be `error.apodiniError.merge(with: authError)`. but I'd consider this less intuitive. I guess it comes down to the term "merge" not specifying a precedence. I'd suggest something like: `authError.caused(by: error.apodiniError)`.

While you're at it, I think we could also make the argument just an error (or at least provide one overload that accepts an `Error`). As shown in the example above, this function will often be used in a `catch`, where the `error` isn't typed.

bauer-andreas replied:
I agree, that making the precedence more explicit makes sense. The merge call wasn't really created with the intention for public use, but yes it is user facing and might be useful anyways. Will go for the `caused(by:)` approach with an Error overload.

bauer-andreas replied:
One note, the `merge` call currently makes sense as we enforce that two Errors which get merged have the same Error type.
Also not sure if the `caused(by:)` really implies that stuff like options and information is actually **merged**.
Typically you would expect (e.g. like Java does it) that it just captures the "previous" error.

So would it maybe be better to keep the `merge` function name, but making the precedence explicit?
E.g. `merge(withRHS:)` or `merge(rhs:)`? (or lhs depending what makes more sense).

theMomax replied:
Good point. Maybe detailed(by:)?

theMomax commented on Sources/ApodiniAuthorization/AuthorizationRequirements/RequirementResult.swift:
```diff
@@ -0,0 +1,51 @@
+    /// When evaluating ``AuthorizationRequirement`` evaluation will continue with the next requirement
+    /// when encountering `undecided`. This is the fundamental difference to the other two cases.
+    /// Although, should the final result of some ``AuthorizationRequirement`` be undecided, it is nonetheless
+    /// and acceptance state. Everything which is not explicitly rejected is accepcted.
```
"an" acceptance state

theMomax commented on Sources/ApodiniVaporSupport/ApodiniError+AbortError.swift:
```diff
@@ -24,38 +25,3 @@ extension ApodiniError: AbortError {
-
-extension HTTPResponseStatus: ApodiniErrorCompliantOption {
-    public static func `default`(for type: ErrorType) -> HTTPResponseStatus {
-        switch type {
```
Was this deleted or moved somewhere else?

bauer-andreas replied:
yes. this was moved into the `ApodiniHTTPProtocol` target, [see](https://github.com/Apodini/Apodini/blob/7ce97bed44164c15c0740818705cbbc2dc0abcad/Sources/ApodiniHTTPProtocol/ApodiniError/HTTPResponseStatusOption.swift).

The `ApdoniHTTPProtocol` is basically for those things which are only SwiftNIO based and don't have a dependence on Vapor. This is e.g. used for the Authentication Schemes which use the error option and the HTTP Information definitions. But the http authentication schemes don't require Vapor.

theMomax commented on Sources/XCTApodini/XCTAssertRuntimeFailure.swift:
```diff
@@ -13,12 +13,12 @@ import XCTest
 /// Asserts that an expression leads to a runtime failure.
 public func XCTAssertRuntimeFailure<T>(
-    _ expression: @escaping @autoclosure () -> T,
+    _ expression: @escaping @autoclosure () throws -> T,
```
I find that a bit misleading, as one could think that throwing is also considered a `RuntimeFailure` and thus result in a fulfilled assertion. Please document accordingly.

theMomax commented on Tests/ApodiniAuthorizationTests/JWTTests.swift:
```diff
@@ -0,0 +1,269 @@
+                Verify(intendedAudience: \.aud, includes: "ExampleAudience")
+                Verify(notBefore: \.nbf)
+
+                Verify(issuer: \.iss, is: "https://other-option.org", "https://example.org")
```
Love it❤️

theMomax commented on Tests/ApodiniAuthorizationTests/JWTTests.swift:
```diff
@@ -0,0 +1,269 @@
+                .typed(String.self)?.content
+        )
+
+        // test fullyCorrectToken
```
I guess that should be `tokenWithoutEmail`?

bauer-andreas replied:
good copy pasta catch 🍝

theMomax reviewed (approved):
Amazing work! I think this is a huge step forward for Apodini👍🏻

I had a few notes and questions along the way. Once they are resolved, you're good to go!

Commit:
Incorporate feedback
